cilium/envoy_wrapper_main.cc

```diff
@@ -0,0 +1,259 @@
+    int rc = 0;
+    int fd_out = -1; // set below when 'rc' is a file descriptor
+
+    switch (msg.hdr.msg_type_) {
+    case TYPE_DUMP_REQUEST: // msg size == header size
+      value_len = dump_capabilities(CAP_EFFECTIVE, buf, buf_size);
+      break;
+    case TYPE_BPF_OPEN_REQUEST: // msg size == header size + path length
+      // zero terminate path name
+      msg.bpf_open_req.path_[msg_len - sizeof(msg.bpf_open_req)] = '\0';
+      {
+	union bpf_attr attr = {};
+	attr.pathname = uintptr_t(msg.bpf_open_req.path_);
+	fd_out = rc = ::syscall(__NR_bpf, BPF_OBJ_GET, &attr, sizeof(attr));
+      }
+      break;
+    case TYPE_BPF_LOOKUP_REQUEST: // key_size = msg_len - sizeof msg.bpf_lookup_req
+      // require at least one byte key
+      if (msg_len < sizeof(msg.bpf_lookup_req) + 1) {
+	fprintf(stderr, "received truncated bpf lookup request (%zd bytes), skipping\n", msg_len);
+	rc = -1;
+	errno = EINVAL;
+	break;
+      }
+      value_len = msg.bpf_lookup_req.value_size_;
+      // Make sure the value fits into available space
+      if (buf_size < value_len) {
+        rc = -1;
+        errno = EINVAL;
+      } else {
+        union bpf_attr attr = {};
+        attr.map_fd = uint32_t(fd_in);
+        attr.key = uintptr_t(msg.bpf_lookup_req.key_);
+        attr.value = uintptr_t(buf);
+        rc = ::syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr));
+      }
+      if (rc != 0) {
+        value_len = 0;
+      }
+      break;
+    case TYPE_SETSOCKOPT32_REQUEST: // msg_len == sizeof msg.setsockopt_req
+      if (msg_len < sizeof(msg.setsockopt_req)) {
+	fprintf(stderr, "received truncated setsockopt request (%zd bytes), skipping\n", msg_len);
+	rc = -1;
+	errno = EINVAL;
+	break;
+      }
+      rc = ::syscall(__NR_setsockopt, fd_in, msg.setsockopt_req.level_,
+		     msg.setsockopt_req.optname_, &msg.setsockopt_req.optval_,
+		     sizeof(msg.setsockopt_req.optval_));
+      break;
+    default:
+      IS_ENVOY_BUG(fmt::format("Unexpected privileged call type: {}", msg.hdr.msg_type_));
+    }
+
+    // Close the received file descriptor
+    if (fd_in != -1) {
+      ::close(fd_in);
+    }
+
+    // Form the response in place
+    msg.response.hdr_.msg_type_ = TYPE_RESPONSE;
+    if (fd_out != -1) {
+      // Pass a poitive but invalid fd in ret_, to be replaced with the passed
+      // fd by the receiver.
+      msg.response.ret_ = {INT_MAX, 0};
+    } else {
+      msg.response.ret_ = {rc, rc != -1 ? 0 : errno};
+    }
+    size = send_fd_msg(&msg, sizeof(msg.response), buf, value_len, fd_out);
+    if (size < ssize_t(sizeof(msg.response) + value_len)) {
+      perror("sendmsg");
+    }
+
+    // Close the sent file descriptor
+    if (fd_out != -1) {
+      ::close(fd_out);
+    }
+  }
+}
+
+} // namespace CiliumEnvoyWrapper
+
+// NOLINT(namespace-envoy)
+
+#define WRAPPER_SUFFIX "-wrapper"
+#define WRAPPER_SUFFIX_LEN (sizeof(WRAPPER_SUFFIX) - 1)
+
+int main(int /* argc */, char** argv) {
```

Comment:
after having a first look at the overall structure (new bazel libs and the `main` of the wrapper) i wonder whether it wouldn't be better to even further separate the wrapper logic from the "privileged service" bits.

Currently:
```mermaid
graph LR;
    W[Wrapper]-- forks -->E[Envoy];
    E-. calls .-> W;
```

Idea:
```mermaid
graph LR;
    W[Wrapper]-- forks -->E[Envoy];
    W[Wrapper]-- forks -->P[Privileged Service];
    E-. calls .-> P;
```

IMO this could lead to better understandable architecture and names (bazel libraries, C++ types, etc). Currently everything is named ~ `EnvoyWrapper**` which is kind of unclear.

* Wrapper: Being responsible to start the necessary processes. Maybe we could even rename this to "EnvoyStarter" to make this more clear.
* Privileged Service: Responsible to provide the privileged actions (BPF map access, socket handling) to the Envoy process. (Same here, we might find a more precise name)

And it might even support cases where the two processes (envoy & "privileged service") are started independent of the wrapper (starter) - e.g. by running the privileged service in a separate sidecar. Or at least we're preparing for this from an architectural point of view.

Reply:
> And it might even support cases where the two processes (envoy & "privileged service") are started independent of the wrapper (starter) - e.g. by running the privileged service in a separate sidecar. Or at least we're preparing for this from an architectural point of view.

This will need a different, more explicit method of passing the fd to the cilium-envoy process. I looked into this briefly, but did not test anything out yet, IMO we can leave this for now. However, if we implement support for the sidecar then the wrapper/starter is not necessary at all. So IMO building more structure into this now would be an overkill.

Better names for the bazel bits are welcomed, I'm pretty bad at naming these kind of things :-)

Reply:
> > And it might even support cases where the two processes (envoy & "privileged service") are started independent of the wrapper (starter) - e.g. by running the privileged service in a separate sidecar. Or at least we're preparing for this from an architectural point of view.
> 
> This will need a different, more explicit method of passing the fd to the cilium-envoy process. I looked into this briefly, but did not test anything out yet, IMO we can leave this for now. However, if we implement support for the sidecar then the wrapper/starter is not necessary at all. So IMO building more structure into this now would be an overkill.
> 
> Better names for the bazel bits are welcomed, I'm pretty bad at naming these kind of things :-)

I agree that the sidecar idea is a bit off for the time being. This shouldn't be the reason.

But as mentioned - it would also help in better understanding the design by separating the responsibilities - and therefore allow for better naming. Naming is especially hard if "the thing" is responsible for multiple tasks. But we might can solve this by splitting up `envoy_wrapper_main.cc` into these two aspects.

Reply:
I'd recommend to separate the main function (which is the actual wrapper / starter) from the "ProtocolServer by moving the ProtocolServer bits into its own file `envoy_wrapper_protocolserver.cc`.

And renaming all `envoy_wrapper_*`  (except `envoy_wrapper_main.cc`)  files to something `privileged_service_*` 

-> 

* `privileged_service_client`
* `privileged_service_server`
* `privileged_service_protocol`

and adapt the bazel envoy libs (and their names accordingly)

Reply:
Renamed and separated as suggested. Also renamed the new binary as `cilium-envoy-starter`, but kept it in a single process that now implements the `PrivilegedServiceServer`.

Reply:
nice - looks great! thanks!